src/ecies.cpp

```diff
@@ -0,0 +1,444 @@
+{
+    return _impl->getEphemeralKey();
+}
+
+std::vector<uint8_t> ECIESEncryptionCtx::getMAC()
+{
+    return _impl->getMAC();
+}
+
+ECIESDecryptionCtx::ECIESDecryptionCtx(
+        const AsymmetricPrivateKey &privKey,
+        const std::shared_ptr<KeyDerivationFunction> kdf,
+        const size_t macKeySize,
+        std::function<std::unique_ptr<MessageAuthenticationCode> (const std::vector<uint8_t> &)> macFactoryFunc,
+        const size_t symCipherKeySize,
+        std::function<std::unique_ptr<SymmetricCipherI>(const std::vector<uint8_t>&,
+                                                        const SymmetricCipherOperation&)> symCipherFactoryFunc,
+        const std::vector<uint8_t> &S1,
+        const std::vector<uint8_t> &S2)
+{
+    _impl = std::make_unique<ECIESDecryptionCtx::Impl>(privKey, kdf, macKeySize, macFactoryFunc,
+                                                       symCipherKeySize, symCipherFactoryFunc, S1, S2);
+}
+
+ECIESDecryptionCtx::~ECIESDecryptionCtx() = default;
+
+void ECIESDecryptionCtx::update(const std::vector<uint8_t> &message)
+{
+    _impl->update(message);
+}
+
+std::vector<uint8_t> ECIESDecryptionCtx::finish()
+{
+    return _impl->finish();
+}
+
+void ECIESDecryptionCtx::setEphemeralKey(const AsymmetricPublicKey &ephKey)
+{
+    _impl->setEphemeralKey(ephKey);
+}
+
+void ECIESDecryptionCtx::setMAC(const std::vector<uint8_t> &tag)
+{
+    _impl->setMAC(tag);
+}
+
+ECIESCtxBuilder::ECIESCtxBuilder()
+{
+
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setKDF(std::shared_ptr<mococrw::KeyDerivationFunction> kdf)
+{
+    _kdf = kdf;
+    return *this;
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setMacCipherKeySize(size_t keySize)
+{
+    _macKeySize = keySize;
+    return *this;
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setMacFactoryFunction(std::function<std::unique_ptr<MessageAuthenticationCode>
+                                                        (const std::vector<uint8_t> &)> func)
+{
+    _macFactoryFunction = func;
+    return *this;
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setSymmetricCipherKeySize(size_t keySize)
+{
+    _symmetricKeySize = keySize;
+    return *this;
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setSymmetricCipherFactoryFunction(std::function<std::unique_ptr<SymmetricCipherI>
+                                                                    (const std::vector<uint8_t> &,
+                                                                     const SymmetricCipherOperation&)> func)
+{
+    _symmetricCipherFactoryFunction = func;
+    return *this;
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setKDFSalt(std::vector<uint8_t> kdfSalt)
+{
+    _kdfSalt = std::move(kdfSalt);
+    return *this;
+}
+
+ECIESCtxBuilder &ECIESCtxBuilder::setMACSalt(std::vector<uint8_t> macSalt)
+{
+    _macSalt = std::move(macSalt);
+    return *this;
+}
+
+ECIESEncryptionCtxBuilder::ECIESEncryptionCtxBuilder(const AsymmetricPublicKey &bobsKey)
+    : _pubKey(bobsKey), _cert(nullptr)
```

Comment:
Please no. The certificate objects were never designed to be like an "optional". No code there is expecting a nullptr and I start wondering about the constructor from pointer being public. This seems in restrospective to be a bad idea. If you want something optional, make it explicitly optional. It's a design flaw in the library that this is possible in the first place.

Reply:
done